### PR TITLE
chore(*) make kumactl install control plane leverage the Helm Charts

### PR DESCRIPTION
**Pull request by nickolaev**
### Summary

Make the `kumactl instal control-plane` generate the manifests based in rendering the Helm Charts.

### Documentation

Internal

### CODE REVIEW

**Comment by nickolaev** [comment]
@austince please check this one

File: app/kumactl/cmd/install/render_helm_files.go [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	}
+	sort.Strings(keys)
+
+	for _, k := range keys {
+		if strings.HasSuffix(k, "yaml") {
+			content := files[k]
+
+			// strip Helm Chart specific labels
+			for _, label := range stripLabels {
+				re := regexp.MustCompile("(?m)[\r\n]+^.*" + label + ".*$")
```
Comment (austince) [inline app/kumactl/cmd/install/render_helm_files.go:155]: nit - might want to compile these all ahead of time instead of for each file.

```go
func labelRegex(label string) *regexp.Regexp {
	return regexp.MustCompile("(?m)[\r\n]+^.*" + label + ".*$")
}

var stripLabelsRegexps = []*regexp.Regexp{
	labelRegex("app.kubernetes.io/managed-by"),
	labelRegex("helm.sh/chart"),
	labelRegex("app.kubernetes.io/version"),
}
```

File: app/kumactl/cmd/install/render_helm_files.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	loadedTemplates := loadedChart.Templates
+	loadedChart.Templates = []*chart.File{}
+
+	for _, t := range loadedTemplates {
+		if !strings.HasPrefix(t.Name, "templates/pre-") {
+			loadedChart.Templates = append(loadedChart.Templates, &chart.File{
+				Name: t.Name,
+				Data: t.Data,
+			})
+		}
```
Comment (austince) [inline app/kumactl/cmd/install/render_helm_files.go:88]: This makes a lot of sense, though we might want to document the `pre-` convention for hooks. What about `post` hooks?

File: app/kumactl/cmd/install/render_helm_files.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	overrideValues := map[string]interface{}{}
+
+	v := reflect.ValueOf(args)
+	t := v.Type()
+	for i := 0; i < t.NumField(); i++ {
+		name := t.Field(i).Name
+		value := v.FieldByName(name).Interface()
+
+		splitName := strings.Split(name, "_")
+		len := len(splitName)
```
Comment (austince) [inline app/kumactl/cmd/install/render_helm_files.go:104]:
```suggestion
		nameLen := len(splitName)
```
nit to avoid conflict w/ `len()` builtin

File: app/kumactl/cmd/install/render_helm_files.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+		for i := 1; i < len-1; i++ {
+			n := splitName[i]
+
+			if _, ok := root[n]; !ok {
+				root[n] = map[string]interface{}{}
+			}
+			root = root[n].(map[string]interface{})
+		}
+		root[splitName[len-1]] = value
```
Comment (austince) [inline app/kumactl/cmd/install/render_helm_files.go:116]: I find this map generation a bit hard to follow. Not surehow difficult something like this would be to implement, but would be nicer to have the arg-to-helm-value mapping tagged on the struct fields instead of using reflection.

```go
type InstallControlPlaneArgs struct {
	controlPlaneImagePullPolicy             string `helm:"controlPlane.image.pullPolicy"`
	// ...
```

File: app/kumactl/cmd/install/render_helm_files.go [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+kind: Namespace
+metadata:
+  name: %s
+  labels:
+    kuma.io/system-namespace: "true"
+`, namespace)
+}
+
+func renderHelmFiles(templates []data.File, args interface{}) ([]data.File, error) {
+	loadedChart, err := loadCharts(templates)
```
Comment (jakubdyszkiewicz) [inline app/kumactl/cmd/install/render_helm_files.go:41]: nit: `loadedChart, err := loadCharts(templates)` -> `chart, err := loadChart(templates)`

File: deployments/charts/kuma/README.md [hunk 6]
```diff
@@ -58,6 +58,10 @@ The chart supports Helm v3+.
 All Kuma CRDs are loaded via the [`crds`](crds) directory. For more detailed information on CRDs and Helm,
 please refer to [the Helm documentation][helm-crd].
 
+## Note to Chart developers
+
+The charts are used internally in `kumactl install`, therefore the following rules apply when developing new chat features:
+ * all templates that start with `pre-` and `post-` are omitted when processing in `kumactl install`
```
Comment (jakubdyszkiewicz) [inline deployments/charts/kuma/README.md:64]: Maybe add a note that you need to run `make generate...` to have consistent HELM and kumactl install?

File: go.mod [hunk 7]
```diff
@@ -59,12 +57,13 @@ require (
 )
 
 replace (
+	github.com/Azure/go-autorest => github.com/Azure/go-autorest v13.3.2+incompatible
```
Comment (jakubdyszkiewicz) [inline go.mod:60]: what was the problem?

File: app/kumactl/cmd/install/testdata/install-control-plane.defaults.golden.yaml [hunk 8]
```diff
@@ -5499,7 +5548,7 @@ webhooks:
           - meshes
     sideEffects: None
   - name: owner-reference.kuma-admission.kuma.io
-    failurePolicy: Fail
+    failurePolicy: Ignore
```
[the remark this answers is not in the document]
Comment (nickolaev) [reply app/kumactl/cmd/install/testdata/install-control-plane.defaults.golden.yaml:5551]: Oh this probably means the Helm charts are wrong.

File: go.mod (hunk 7 above)
Comment (nickolaev) [reply go.mod:60]: I don't remember anymore, but some conflict in the versions. In any case, this was borrowed from Helm itself -> https://github.com/helm/helm/blob/master/go.mod#L49

File: deployments/charts/kuma/README.md (hunk 6 above)
Comment (nickolaev) [reply deployments/charts/kuma/README.md:64]: Makes sense, but I think there is still this verification part of `make test`

**Review by jakubdyszkiewicz — APPROVED** [review]

**Review by austince — APPROVED** [review]
looks great to me too!
